### PR TITLE
[SYCLomatic][part4] Refine cmake migration framework to make Priority field be consistent with that of other user defined rules

### PR DESCRIPTION
**Pull request by tomflinda**
1. Add new field "CmakeSyntax"  to identify cmake syntax need to be migrated.
2. Add new kind value "CMakeRule" to identify user defined yaml rules.

### CODE REVIEW

File: clang/lib/DPCT/MigrateCmakeScript.cpp [hunk 1]
```diff
@@ -656,11 +667,42 @@ static void storeBufferToFile() {
...
+void registerCmakeMigrationRule(MetaRuleObject &R) {
+  auto PR =
+      MetaRuleObject::PatternRewriter(R.In, R.Out, R.Subrules, R.MatchMode,
+                                      R.RuleId, R.CmakeSyntax, R.Priority);
+
+  auto Iter = CmakeBuildInRules.find(PR.CmakeSyntax);
+  if (Iter != CmakeBuildInRules.end()) {
+    if (PR.Priority == RulePriority::Takeover) {
+      assert(Iter->second.Priority < PR.Priority &&
+             "Two same cmake synaxes have \'Takeover\' priority.\n ");
```
Comment (zhiweij1) [inline clang/lib/DPCT/MigrateCmakeScript.cpp]:
```suggestion
             "Two same cmake syntaxes have \'Takeover\' priority.\n ");
```

File: clang/lib/DPCT/MigrateCmakeScript.cpp [hunk 2]
```diff
@@ -594,26 +602,29 @@ static void unifyInputFileFormat() {
 }
 
 static void applyCmakeMigrationRules() {
+
+  std::map<std::string, void (*)(std::string &, size_t &, size_t &)>
```
Comment (zhiweij1) [inline clang/lib/DPCT/MigrateCmakeScript.cpp]:
```suggestion
  static const std::map<std::string, void (*)(std::string &, size_t &, size_t &)>
```

Comment (tomflinda) [reply clang/lib/DPCT/MigrateCmakeScript.cpp]: Okay, will update it.

File: clang/lib/DPCT/MigrateCmakeScript.cpp (hunk 1 above)
Comment (tomflinda) [reply clang/lib/DPCT/MigrateCmakeScript.cpp]: Will fix it in the updated PR>

**Review by AndyCHHuang — APPROVED** [review]
LGTM

**Review by zhiweij1 — APPROVED** [review]

**Review by ziranzha — APPROVED** [review]

File: clang/lib/DPCT/MigrateCmakeScript.cpp [hunk 3]
```diff
@@ -594,26 +602,30 @@ static void unifyInputFileFormat() {
...
-      // emitted to notify the user. Currently, only migration of
-      // cmake_minimum_required() is implemented by implicit migration rule.
-      if (PR.RuleId.find("cmake_minimum_required") != std::string::npos) {
-        llvm::outs() << "Migration for cmake_minimum_required has been "
-                        "implemented as implicit rule, so migration rule \'"
-                     << PR.RuleId << "\' is skipped.\n ";
-        continue;
+    for (const auto &CmakeSyntaxEntry : CmakeBuildInRules) {
+      const auto &PR = CmakeSyntaxEntry.second;
+      if (PR.In.empty() && PR.Out.empty()) {
```
Comment (zhimingwang36) [inline clang/lib/DPCT/MigrateCmakeScript.cpp:618]: please refine L618,  don't assume in/out is always not empty.

**Pull request merged**